File: app/src/main/java/com/kota/Telnet/Reference/TelnetKeyboard.java
package com.kota.Telnet.Reference;

public class TelnetKeyboard {
    public static final int DELETE = 265;
    public static final int DOWN_ARROW = 259;
    public static final int END = 263;
    public static final int HOME = 262;
    public static final int INSERT = 264;
    public static final int LEFT_ARROW = 256;
    public static final int PAGE_DOWN = 261;
    public static final int PAGE_UP = 260;
    public static final int RIGHT_ARROW = 257;
    public static final int SPACE = 32;
    public static final int TAB = 9;
    public static final int UP_ARROW = 258;
    public static final int CTRL_G = 7;
    public static final int CTRL_P = 16;
    public static final int CTRL_Q = 17;
    public static final int CTRL_R = 18;
    public static final int CTRL_S = 19;
    public static final int CTRL_U = 21;
    public static final int CTRL_X = 24;
    public static final int CTRL_Y = 25;
    public static final int BACK_ONE_CHAR = 83;
    public static final int SHIFT_M = 115;

    public static byte[] getKeyDataWithTimes(int keyCode, int times) {
        byte[] keydata = getKeyData(keyCode);
        byte[] data = new byte[(keydata.length * times)];
        for (int index = 0; index < data.length; index++) {
            data[index] = keydata[index % keydata.length];
        }
        return data;
    }

    /**
     * 輸入鍵盤指令的代號，回傳 telnet command
     * @param keyCode 鍵盤指令
     * @return byte[] telnet command
     * */
    public static byte[] getKeyData(int keyCode) {
        switch (keyCode) {
            case TAB: /* tab */
                return new byte[]{9};
            case SPACE: /* space */
                return new byte[]{32};
            case LEFT_ARROW: /* Left Arrow */
                return new byte[]{27, 91, 68};
            case RIGHT_ARROW: /* Right Arrow */
                return new byte[]{27, 91, 67};
            case UP_ARROW /*258*/:
                return new byte[]{27, 91, 65};
            case DOWN_ARROW /*259*/:
                return new byte[]{27, 91, 66};
            case PAGE_UP /*260*/:
                return new byte[]{27, 91, 53, 126};
            case PAGE_DOWN /*261*/:
                return new byte[]{27, 91, 54, 126};
            case HOME /*262*/:
                return new byte[]{27, 91, 49, 126};
            case END /*263*/:
                return new byte[]{27, 91, 52, 126};
            case INSERT /*264*/:
                return new byte[]{27, 91, 50, 126};
            case DELETE /*265*/:
                return new byte[]{27, 91, 51, 126};
            default:
                return new byte[]{(byte) keyCode};
        }
    }
}

File: app/src/main/java/com/kota/telnet/reference/TelnetKeyboard.kt
package com.kota.telnet.reference

object TelnetKeyboard {
    const val DELETE: Int = 265
    const val DOWN_ARROW: Int = 259
    const val END: Int = 263
    const val HOME: Int = 262
    const val INSERT: Int = 264
    const val LEFT_ARROW: Int = 256
    const val PAGE_DOWN: Int = 261
    const val PAGE_UP: Int = 260
    const val RIGHT_ARROW: Int = 257
    const val SPACE: Int = 32
    const val TAB: Int = 9
    const val UP_ARROW: Int = 258
    const val CTRL_G: Int = 7
    const val CTRL_P: Int = 16
    const val CTRL_Q: Int = 17
    const val CTRL_R: Int = 18
    const val CTRL_S: Int = 19
    const val CTRL_U: Int = 21
    const val CTRL_X: Int = 24
    const val CTRL_Y: Int = 25
    const val BACK_ONE_CHAR: Int = 83
    const val KEY_S: Int = 115
    const val SHIFT_M: Int = 77
    const val small_C: Int = 99

    fun getKeyDataWithTimes(keyCode: Int, times: Int): ByteArray {
        val keyData = getKeyData(keyCode)
        val data = ByteArray((keyData.size * times))
        for (index in data.indices) {
            data[index] = keyData[index % keyData.size]
        }
        return data
    }

    /**
     * 輸入鍵盤指令的代號，回傳 telnet command
     * @param keyCode 鍵盤指令
     * @return byte[] telnet command
     */
    @JvmStatic
    fun getKeyData(keyCode: Int): ByteArray {
        return when (keyCode) {
            TAB -> byteArrayOf(9)
            SPACE -> byteArrayOf(32)
            LEFT_ARROW -> byteArrayOf(27, 91, 68)
            RIGHT_ARROW -> byteArrayOf(27, 91, 67)
            UP_ARROW -> byteArrayOf(27, 91, 65)
            DOWN_ARROW -> byteArrayOf(27, 91, 66)
            PAGE_UP -> byteArrayOf(27, 91, 53, 126)
            PAGE_DOWN -> byteArrayOf(27, 91, 54, 126)
            HOME -> byteArrayOf(27, 91, 49, 126)
            END -> byteArrayOf(27, 91, 52, 126)
            INSERT -> byteArrayOf(27, 91, 50, 126)
            DELETE -> byteArrayOf(27, 91, 51, 126)
            else -> byteArrayOf(keyCode.toByte())
        }
    }
}
